File: frontend/lib/types/practice.ts
import { z } from 'zod';

export const multipleChoiceOption = z.object({
  description: z.string(),
  isCorrect: z.boolean(),
});

export type MultipleChoiceOption = z.infer<typeof multipleChoiceOption>;

export const practiceQuestion = z.object({
  description: z.string(),
  options: z.array(multipleChoiceOption),
});

export type PracticeQuestion = z.infer<typeof practiceQuestion>;

export const practiceQuestions = z.object({
  questions: z.array(practiceQuestion),
});

export type PracticeQuestions = z.infer<typeof practiceQuestions>;

File: server/app/models/practice.py
from pydantic import BaseModel, Field


class MultipleChoiceOption(BaseModel):
    description: str = Field(description="The description of the option")
    isCorrect: bool = Field(description="Whether the option is correct")


class PracticeQuestion(BaseModel):
    description: str = Field(description="The problem description")
    options: list[MultipleChoiceOption] = Field(
        description="2-4 relevant multiple choice options for the problem description. Only one option should be correct"
    )
    explanation: str = Field(
        description="The explanation as to why the option is correct and others are wrong"
    )


class PracticeQuestions(BaseModel):
    questions: list[PracticeQuestion] = Field(description="The list of practice questions")
